### PR TITLE
[amdgpu] Pass variadic arguments without splitting

### PR DESCRIPTION
**Pull request by JonChesterfield**
Pass variadic arguments without changing their type, unlike the fixed ones.

Fixed arguments are modified to better fit into registers. This patch leaves those unchanged.

Splitting struct types into individual fields and packing small structs into integers works well for passing via registers. Variadic arguments are currently unimplemented in the backend. They're likely to be implemented as a pointer to stack memory in which case register-themed optimisations are inapplicable.

Splitting the struct into fields makes it difficult to implement va_arg robustly. The rules around padding and alignment to inverse the struct splitting could be constructed, but at high complexity and no particular advantage.

Passing types as-is means there is a 1:1 correspondence with the type information va_arg has to work with and the parameter type at the call site.

This is an ABI change, but as the only functions affected are variadic ones which are presently a compilation error, not a functional break. Factored out of the larger #93362 and can land independently.

### CODE REVIEW

File: clang/test/CodeGen/amdgpu-variadic-call.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+// CHECK-NEXT:    tail call void (i32, ...) @sink_1(i32 noundef [[F0]], [[STRUCT_PAIR_F64]] [[DOTFCA_1_INSERT]], double noundef [[CONV]], [[UNION_UNION_F32_I32]] [[DOTFCA_0_INSERT]], i32 noundef [[V3]]) #[[ATTR2]]
+// CHECK-NEXT:    tail call void (double, i32, ...) @sink_2(double noundef [[F1]], i32 noundef [[F0]], [[STRUCT_PAIR_F64]] [[DOTFCA_1_INSERT]], double noundef [[CONV]], [[UNION_UNION_F32_I32]] [[DOTFCA_0_INSERT]], i32 noundef [[V3]]) #[[ATTR2]]
+// CHECK-NEXT:    ret void
+//
+void multiple_two(int f0, double f1, pair_f64 v0, float v1, union_f32_i32 v2, int v3)
+{
+  sink_0(v0, v1, v2, v3);
+  sink_1(f0, v0, v1, v2, v3);
+  sink_2(f1, f0, v0, v1, v2, v3);
+}
```
Comment (arsenm) [inline clang/test/CodeGen/amdgpu-variadic-call.c:293]: Also test some structs that are so big they wouldn't be passed directly?

File: clang/lib/CodeGen/Targets/AMDGPU.cpp [hunk 2]
```diff
@@ -197,12 +202,20 @@ ABIArgInfo AMDGPUABIInfo::classifyKernelArgumentType(QualType Ty) const {
...
 }
 
-ABIArgInfo AMDGPUABIInfo::classifyArgumentType(QualType Ty,
+ABIArgInfo AMDGPUABIInfo::classifyArgumentType(QualType Ty, bool Variadic,
                                                unsigned &NumRegsLeft) const {
   assert(NumRegsLeft <= MaxNumRegsForArgsRet && "register estimate underflow");
 
   Ty = useFirstFieldIfTransparentUnion(Ty);
 
+  if (Variadic) {
```
Comment (arsenm) [inline clang/lib/CodeGen/Targets/AMDGPU.cpp:211]: Wouldn't we still want to follow the isAggregateTypeForABI rules? Large structs should still go through byref?

Comment (arsenm) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:211]: The bigger concern is using giant aggregate types is not great IR, and not all that well supported. a 65k element array will crash in SelectionDAG for example

**Comment by JonChesterfield** [comment]
That's our call really. Right now this passes everything as whatever type it claimed to be as far as creating the call instruction goes, then variadic lowering pastes them all into a single structure with four byte alignment on every field.

Tagging some parameters as byval or byref is fine as far as variadic lowering goes. They'll tend to end up in their own alloca, then a ptr in the call frame. It means more test variation but the common path knows what byval is, and adding byref will be fine.

I'm not immediately worried about sdag crashing if people pass loads of data to a variadic call, considering our baseline capability here is no variadics at all. Likewise I'm not sure what the performance tradeoffs of copying into the frame vs indirect reference are and not very worried about making an optimal choice there.

I'll add some more structs to show they're also passed unchanged. If you want more complicated rules you're welcome to them - happy to pass arguments in whatever pattern you like, though the splitting them into multiple fields as we currently do would push the va_arg lowering complexity up alarmingly.

What would you sign off on? The current approach is the simplest. Next up is using the same when-to-pass-byref as fixed arguments without the splitting. Otherwise we can have some other heuristic.

**Comment by JonChesterfield** [comment]
Slightly jumped the gun on this patch because Eli just wrote a bunch of useful things on #93362 and there's therefore a credible chance that rebasing on this will get the pass in the hands of other people soon, giving libc sprintf and so forth.

**Comment by arsenm** [comment]
> @arsenm You're right about passing larger things indirectly. I'm intending to land this as-is, with the types inlined, as that unblocks #93362. I'm nervous that the extra pointer indirection will hit the same memory error that tweaking codegen in that patch hits (it's a similar sort of pattern to the top level argument passing) and wish to postpone that until there is a working baseline.

I do think we need to revisit some threshold at some point. We should use byref for anything that's most likely to hit the stack anyway

File: clang/test/CodeGen/amdgpu-variadic-call.c (hunk 1 above)
Comment (arsenm) [reply clang/test/CodeGen/amdgpu-variadic-call.c:293]: Missing big tests?

**Pull request merged**